src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+
+            var masked = (lenByte & 0x01) != 0;
+            var payloadLen = (lenByte & 0xFE) >> 1;
+
+            // Mark what we've got so far as consumed
+            _inbound.Advance(buffer.Start);
+
+            // Calculate the rest of the header length
+            var headerLength = masked ? 4 : 0;
+            if (payloadLen == 126)
+            {
+                headerLength += 2;
+            }
+            else if (payloadLen == 127)
+            {
+                headerLength += 4;
+            }
+
+            var maskingKey = default(ReadableBuffer);
+            if (headerLength > 0)
+            {
+                buffer = await _inbound.ReadAtLeastAsync(headerLength, cancellationToken);
+
+                // Read extended payload length (if any)
+                if (payloadLen == 126)
+                {
+                    payloadLen = buffer.ReadBigEndian<ushort>();
+                    buffer = buffer.Slice(sizeof(ushort));
+                }
+                else if (payloadLen == 127)
+                {
+                    var longLen = buffer.ReadBigEndian<ulong>();
+                    buffer = buffer.Slice(sizeof(ulong));
+                    if (longLen > int.MaxValue)
+                    {
+                        throw new WebSocketException($"Frame is too large. Maximum frame size is {int.MaxValue} bytes");
+                    }
+                    payloadLen = (int)longLen;
+                }
+
+                // Read masking key
+                if (masked)
+                {
+                    var maskingKeyStart = buffer.Start;
+                    maskingKey = buffer.Slice(0, 4);
+
+                    // Mark the masking key examined but not consumed
+                    _inbound.Advance(maskingKeyStart, buffer.Start);
+                }
+                else
+                {
+                    // Mark everything consumed, there was no masking key
+                    _inbound.Advance(buffer.Start);
+                }
+            }
+
+            var payload = default(ReadableBuffer);
+            if (payloadLen > 0)
+            {
+                buffer = await _inbound.ReadAtLeastAsync(payloadLen, cancellationToken);
+
+                if(masked)
+                {
+                    // Need to skip the masking key, since we didn't consume it.
+                    buffer = buffer.Slice(4);
+                }
+                payload = buffer.Slice(0, payloadLen);
+
+                // We don't slice the buffer because we actually want to keep the payload marked unconsumed until the next read
+                // Instead, we capture the position after the payload as the "lastFrameCursor"
+                _lastPayloadEnd = buffer.Slice(payloadLen).Start;
+
+                if (masked)
+                {
+                    unsafe
+                    {
+                        // TODO: Unmask in place? Can we do that safely? If so we can avoid a heap allocation here.
```

Comment:
you should be fine to do this *as long as* you're not using `Append(ReadableBuffer)` with something that you didn't expect to change; in all other cases, you are (IIRC) looking at isolated data